rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -2344,6 +2344,100 @@ impl SystemState {
+            memory_usage,
+            wasm_memory_usage,
+            memory_allocation,
+            wasm_memory_limit,
+            wasm_memory_threshold,
+        ) {
+            self.task_queue.enqueue(ExecutionTask::OnLowWasmMemory);
+        } else {
+            self.task_queue.remove(ExecutionTask::OnLowWasmMemory);
+        }
+    }
+}
+
+/// Condition for `OnLowWasmMemoryHook` is satisfied if the following holds:
+///
+/// 1. In the case of `memory_allocation`
+///     `wasm_memory_threshold >= min(memory_allocation - memory_usage_without_wasm_memory, wasm_memory_limit) - wasm_memory_usage`
+/// 2. Without memory allocation
+///     `wasm_memory_threshold >= wasm_memory_limit - wasm_memory_usage`
+///
+/// Note: if `wasm_memory_limit` is not set, its default value is 4 GiB.
+pub fn is_low_wasm_memory_hook_condition_satisfied(
+    memory_usage: NumBytes,
+    wasm_memory_usage: NumBytes,
+    memory_allocation: Option<NumBytes>,
+    wasm_memory_limit: Option<NumBytes>,
+    wasm_memory_threshold: NumBytes,
+) -> bool {
+    // If wasm memory limit is not set, the default is 4 GiB. Wasm memory
+    // limit is ignored for query methods, response callback handlers,
+    // global timers, heartbeats, and canister pre_upgrade.
+    let wasm_memory_limit =
+        wasm_memory_limit.unwrap_or_else(|| NumBytes::new(4 * 1024 * 1024 * 1024));
+
+    debug_assert!(
+        wasm_memory_usage <= memory_usage,
+        "Wasm memory usage {} is greater that memory usage {}.",
+        wasm_memory_usage,
+        memory_usage
+    );
+
+    let memory_usage_without_wasm_memory =
+        NumBytes::new(memory_usage.get().saturating_sub(wasm_memory_usage.get()));
+
+    // If the canister has memory allocation, then it maximum allowed Wasm memory can be calculated
+    // as min(memory_allocation - memory_usage_without_wasm_memory, wasm_memory_limit).
+    let wasm_capacity = memory_allocation.map_or_else(
+        || wasm_memory_limit,
+        |memory_allocation| {
+            debug_assert!(
+                memory_usage_without_wasm_memory <= memory_allocation,
+                "Used non-Wasm memory: {:?} is larger than memory allocation: {:?}.",
+                memory_usage_without_wasm_memory,
+                memory_allocation
+            );
+            std::cmp::min(
+                memory_allocation - memory_usage_without_wasm_memory,
+                wasm_memory_limit,
+            )
+        },
+    );
+
+    // Conceptually we can think that the remaining Wasm memory is
+    // equal to `wasm_capacity - wasm_memory_usage` and that should
+    // be compared with `wasm_memory_threshold` when checking for
+    // the condition for the hook. However, since `wasm_memory_limit`
+    // is ignored in some executions as stated above it is possible
+    // that `wasm_memory_usage` is greater than `wasm_capacity` to
+    // avoid overflowing subtraction we adopted inequality.
+    wasm_capacity < wasm_memory_usage + wasm_memory_threshold
```

Comment:
This whole function is moved from `sandbox_safe_system_state` to here to avoid having two copies.

Reply:
> You are right that is not easy possible

Approving as is to unblock you, but what's the issue with having `update_on_low_wasm_memory_hook_status()` somewhere next to `process_install_code_result()` and keeping `is_low_wasm_memory_hook_condition_satisfied()` in `sandbox_safe_system_state`, where it was?

All you need is a few bits of information from `SystemState` (`memory_allocation`, `wasm_memory_limit` and `wasm_memory_threshold`), then call `is_low_wasm_memory_hook_condition_satisfied()` with them. Then enqueue or remove the task as necessary. All from within `execution_environment`.

Reply:
I will merge it as is because I am taking off until the end of the week, and we can continue discussing it here. If we move `is_low_wasm_memory_hook_condition_satisfied` to `system_api(sandbox_safe_system_state)`, then we will need `replicated_state` to depend on `system_api` (to use `is_low_wasm_memory_hook_condition_satisfied` on `system_state`), but since `system_api` already depends on `replicated_state` that will create a cyclic dependency. Am I missing something?

Reply:
> then we will need `replicated_state` to depend on `system_api` (to use `is_low_wasm_memory_hook_condition_satisfied` on `system_state`)

You won't if you don't call `is_low_wasm_memory_hook_condition_satisfied` from `system_state`. That's what I've been trying to say.

Just move `update_on_low_wasm_memory_hook_status()` out of `system_state`, since it's also Execution-specific logic and all it needs is three numbers out of `SystemState` (as opposed to e.g. needing mutable access to private `SystemState` fields).

Your interaction with `SystemState` should be limited to calling `TaskQueue::enqueue()` and `TaskQueue::remove()` and retrieving memory limits/thresholds/allocations from it.